conf.c

```diff
@@ -93,168 +128,482 @@ chomp(char *str)
-                                user = NULL;
-			config.masquerade_host = host;
-			config.masquerade_user = user;
-		} else if (strcmp(word, "STARTTLS") == 0 && data == NULL)
-			config.features |= STARTTLS;
-		else if (strcmp(word, "FINGERPRINT") == 0) {
-			if (strlen(data) != SHA256_DIGEST_LENGTH * 2) {
-				errlogx(EX_CONFIG, "invalid sha256 fingerprint length");
-			}
-			unsigned char *fingerprint = malloc(SHA256_DIGEST_LENGTH);
-			if (fingerprint == NULL) {
-				errlogx(EX_CONFIG, "fingerprint allocation failed");
-			}
-			for (unsigned int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
-				if(sscanf(data + 2 * i, "%02hhx", &fingerprint[i]) != 1) {
-					errlogx(EX_CONFIG, "failed to read fingerprint");
+			if(item->checkFunction != NULL)
+				if(item->checkFunction(identifier, value) != 0)
+					return EX_CONFIG;
+
+			/* else let's set the config flag to on
+			 * we do not bother changing the value if it is already set,
+			 * as its concrete value is meaningless.
+			 * The value can already be set,
+			 * if its flag has been enabled more than once in the config file. */
+			if(item->str_value == NULL)
+				item->str_value = strdup("ON");
+			return 0;
+		}
+	}
+	else /* Item has not been found */
+		return EX_CONFIG;
+}
+
+void free_all_configuration_settings(void)
+{
+	struct config_item_t* current_item;
+	struct config_item_t* next_item;
+
+	last_checked_item = NULL;
+
+	current_item = SLIST_FIRST(&config_head);
+	while (current_item != NULL)
+	{
+		next_item = SLIST_NEXT(current_item, next_item);
+
+		if(current_item->identifier != NULL)
+			free(current_item->identifier);
+		if(current_item->str_value != NULL)
+			free(current_item->str_value);
+
+		free(current_item);
+		current_item = next_item;
+	}
+	SLIST_INIT(&config_head);
+}
+
+static int check_fingerprint_configuration(const char *identifier, const char *value)
+{
+	unsigned int counter;
+	unsigned char *fingerprint;
+
+	if(identifier == NULL || value == NULL)
+		return EX_CONFIG;
+
+	if(strlen(value) != SHA256_DIGEST_LENGTH * 2)
+		return EX_CONFIG;
+
+	fingerprint = malloc(SHA256_DIGEST_LENGTH);
+	if (fingerprint == NULL)
+		return EX_OSERR;
+
+	for (counter = 0; counter < SHA256_DIGEST_LENGTH; counter++)
+		if(sscanf(value + 2 * counter, "%02hhx", &fingerprint[counter]) != 1)
+		{
+			free(fingerprint);
+			return EX_CONFIG;
+		}
+
+	free(fingerprint);
+	return 0;
+}
+
+/* Helper function to return the masquerade settings split up in login and host */
+struct masquerade_config_t *extract_masquerade_settings(const char *value)
+{
+	char *user;
+	char *host;
+	char *copy_of_value;
+
+	struct masquerade_config_t *masquerade;
+
+#ifdef DEBUG_CONF
+	printf("extract_masquerade_settings called with value '%s'\n", value);
+#endif
+	copy_of_value = host = user = NULL;
+
+	if(value == NULL)
+		return NULL;
+
+	masquerade = malloc(sizeof(struct masquerade_config_t));
+
+	copy_of_value = strdup(value);
+
+	if(masquerade == NULL || copy_of_value == NULL)
```

Comment:
could leak memory, but i guess we would error out hard right after?

Reply:
You're right, that was lazy. I've added a check and hard-exit in the calling function.